### ISSUE TITLE
o or O in command mode cause exception in cshtml file

### ISSUE THREAD

**Issue by ThatRendle** [VsVim/VsVim#946]
When editing a cshtml file, pressing o or O to insert a new line causes an error when it's the first time in a session. Subsequent presses insert 3 lines and don't go into Edit mode.

Environment: VS2012 RC on Windows 8 x64 RP.
Other editor extensions: ReSharper 7 RTW, GhostDoc, Mindscape Web Workbench, nCrunch, Web Tooling Extensions.

Exception from ActivityLog.xml:

```
System.ArgumentException: The supplied SnapshotPoint is on an incorrect snapshot. 
Parameter name: bufferPosition 
at Microsoft.VisualStudio.Text.Editor.Implementation.CaretElement.InternalMoveTo(VirtualSnapshotPoint bufferPosition, PositionAffinity caretAffinity, Boolean captureHorizontalPosition, Boolean captureVerticalPosition, Boolean raiseEvent) 
at Microsoft.VisualStudio.Text.Editor.Implementation.CaretElement.MoveTo(VirtualSnapshotPoint bufferPosition) 
at Vim.TextViewUtil.MoveCaretToVirtualPointRaw[a](ITextView textView, VirtualSnapshotPoint point, a flags) 
at Vim.CommandUtil.InsertLineBelow(Int32 count) 
at Vim.CommandUtil.RunNormalCommand(NormalCommand command, CommandData data) 
at Vim.CommandRunner.Run(KeyInput ki) 
at Vim.CommandRunner.Vim-ICommandRunner-Run(KeyInput ki) 
at Vim.Modes.Normal.NormalMode.ProcessCore(KeyInput ki) 
at Vim.Modes.Normal.NormalMode.Vim-IMode-Process(KeyInput ki) 
at Vim.VimBuffer.ProcessOneKeyInput(KeyInput keyInput) 
at <StartupCode$Vim-Core>.$VimBuffer.processSet@386(VimBuffer x, FSharpRef`1 processResult, KeyInputSet keyInputSet) 
at Vim.VimBuffer.ProcessCore(KeyInputSet keyInputSet) 
at Vim.VimBuffer.Process(KeyInput keyInput) 
at Vim.VimBuffer.Vim-IVimBuffer-Process(KeyInput keyInput) 
at Vim.UI.Wpf.VimKeyProcessor.TryProcess(KeyInput keyInput) 
at VsVim.VsKeyProcessor.TryProcess(KeyInput keyInput) 
at Vim.UI.Wpf.VimKeyProcessor.KeyDown(KeyEventArgs args) 
at Microsoft.VisualStudio.Text.Editor.Implementation.KeyProcessorDispatcher.<DispatchKeyEvents>b__c(KeyProcessor p, KeyEventArgs args) 
at Microsoft.VisualStudio.Text.Editor.Implementation.KeyProcessorDispatcher.<>c__DisplayClass1e`1.<Dispatch>b__1a() 
at Microsoft.VisualStudio.Text.Utilities.GuardedOperations.CallExtensionPoint(Object errorSource, Action call)
```

**Comment by l3m** [comment]
I have had this exception thrown multiple times within a short time in VS2015 with VsVim.

**Comment by jaredpar** [comment]
@l3m have you seen the exception with the <kbd>o</kbd> or another command?

**Comment by l3m** [comment]
It happened quite a few times. I tested it now and could only reproduce it with o, but I am quite certain it also happened with i. 


Also, here's the activity log entry, if that helps: 

<entry>
    <record>739</record>
    <time>2015/07/27 09:17:51.511</time>
    <type>Error</type>
    <source>Editor or Editor Extension</source>
    <description>System.ArgumentException: The supplied SnapshotPoint is on an incorrect snapshot.&#x000D;&#x000A;Parameter name: bufferPosition&#x000D;&#x000A;   at Microsoft.VisualStudio.Text.Editor.Implementation.CaretElement.InternalMoveTo(VirtualSnapshotPoint bufferPosition, PositionAffinity caretAffinity, Boolean captureHorizontalPosition, Boolean captureVerticalPosition, Boolean raiseEvent)&#x000D;&#x000A;   at Microsoft.VisualStudio.Text.Editor.Implementation.CaretElement.MoveTo(SnapshotPoint bufferPosition)&#x000D;&#x000A;   at Vim.TextViewUtil.MoveCaretToPointRaw[a](ITextView textView, SnapshotPoint point, a flags)&#x000D;&#x000A;   at Vim.CommonOperations.MoveCaretToPoint(SnapshotPoint point, ViewFlags viewFlags)&#x000D;&#x000A;   at &lt;StartupCode$Vim-Core&gt;.$CommandUtil.DeleteCharacterAtCaret@596.Invoke(Unit unitVar0)&#x000D;&#x000A;   at Vim.UndoRedoOperations.EditWithUndoTransaction[b](String name, ITextView textView, FSharpFunc`2 action)&#x000D;&#x000A;   at Vim.UndoRedoOperations.Vim-IUndoRedoOperations-EditWithUndoTransaction[a](String name, ITextView textView, FSharpFunc`2 action)&#x000D;&#x000A;   at Vim.CommandUtil.DeleteCharacterAtCaret(Int32 count, Register register)&#x000D;&#x000A;   at Vim.CommandUtil.RunNormalCommand(NormalCommand command, CommandData data)&#x000D;&#x000A;   at Vim.CommandRunner.Run(KeyInput ki)&#x000D;&#x000A;   at Vim.CommandRunner.Vim-ICommandRunner-Run(KeyInput ki)&#x000D;&#x000A;   at Vim.Modes.Normal.NormalMode.Process(KeyInput keyInput)&#x000D;&#x000A;   at Vim.Modes.Normal.NormalMode.Vim-IMode-Process(KeyInput keyInput)&#x000D;&#x000A;   at Vim.VimBuffer.ProcessOneKeyInput(KeyInput keyInput)&#x000D;&#x000A;   at &lt;StartupCode$Vim-Core&gt;.$VimBuffer.processSet@447(VimBuffer x, FSharpRef`1 processResult, KeyInputSet keyInputSet)&#x000D;&#x000A;   at Vim.VimBuffer.ProcessCore(KeyInputSet keyInputSet)&#x000D;&#x000A;   at Vim.VimBuffer.Process(KeyInput keyInput)&#x000D;&#x000A;   at Vim.VimBuffer.Vim-IVimBuffer-Process(KeyInput keyInput)&#x000D;&#x000A;   at Vim.UI.Wpf.VimKeyProcessor.TryProcess(KeyInput keyInput)&#x000D;&#x000A;   at Vim.VisualStudio.Implementation.Misc.VsVimKeyProcessor.TryProcess(KeyInput keyInput)&#x000D;&#x000A;   at Vim.UI.Wpf.VimKeyProcessor.TextInput(TextCompositionEventArgs args)&#x000D;&#x000A;   at Microsoft.VisualStudio.Text.Editor.Implementation.KeyProcessorDispatcher.&lt;&gt;c.&lt;DispatchTextInputEvents&gt;b__6_0(KeyProcessor p, TextCompositionEventArgs args)&#x000D;&#x000A;   at Microsoft.VisualStudio.Text.Editor.Implementation.KeyProcessorDispatcher.&lt;&gt;c__DisplayClass10_1`1.&lt;Dispatch&gt;b__0()&#x000D;&#x000A;   at Microsoft.VisualStudio.Text.Utilities.GuardedOperations.CallExtensionPoint(Object errorSource, Action call)</description>

**Comment by jaredpar** [comment]
@l3m thanks for the entry log.  Unfortunately it confuses things a bit as that code path has been hardened against this particular type of error (other extensions doing post edit cleanups).  Also this stack is pointing to <kbd>x</kbd> or <kbd>Del</kbd> as being the culprit.  Possible though they are hitting a common code path with <kbd>o</kbd>. 

Can you tell me a bit more about your setup?  In particular what other extensions you have installed and what project type / file you commonly see this in.

**Comment by l3m** [comment]
Mainly on a C#-project, so while editing .cs files. 

VS2015, I have ReSharper Ultimate 9.1.3, VsVim 2.0.1 and Viasfora 2.5.62 installed, on Windows 7 Enterprise x64.

**Comment by l3m** [comment]
Hm, it looks like it might be related to ReSharper. I updated to ReSharper from 9.1.something (not sure) to 9.1.3 yesterday, and since then, the crash did not reoccur. 

There was also some weird interaction between the non-updated ReSharper and VsVim, which also did not reoccur.

**Comment by jaredpar** [comment]
@l3m ah okay.  It's possible there was an interaction with the other issue R# fixed from 9.1.2.  If it shows up again with 9.1.3 let me know and I'll take a look.
